cmd/clusterctl/pkg/client/alias.go

```diff
@@ -0,0 +1,27 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/pkg/client/config"
+)
+
+// Alias creates local alias for types defined in the low-level library.
+// By using local alias, ensures that the users of the API will be forced to import the clusterctl high-level library only.
```

Comment:
Is this a common pattern in go? 
Is there a reason for not putting types like Provider into a separate types/api package?

Reply:
> Is this a common pattern in go?

For sure there are other projects using this pattern (e.g kubebuilder)

But, just to make this clear, I think the alias is something nice to have from a consumer point of view, but it less important than many other things. So if this is something blocking/hard to get agreement on, I have no objecting to drop them.

> Is there a reason for not putting types like Provider into a separate types/api package?

If you put types in a separate package, then the user will be forced to import both types and the client package, and this has the same disadvantages of making the users import the low-level library packages and the client package.

Reply:
If we want to ensure users can't import something it should live in `internal` instead of redefining the type. I don't quite follow the point you're making about putting types in a separate package forcing users to import both, do you have an example of the problem this code is trying to solve?

Reply:
As far as I understood, the alias pattern here follows what controller runtime is doing today. Provides an easier facade for users in the main package, but you can still use the inner ones if folks want to access more resources

Reply:
Reference: https://github.com/kubernetes-sigs/controller-runtime/blob/master/alias.go

Reply:
Thanks for pointing that out. 
:+1: